src/zino/trapobservers/link_traps.py

```diff
@@ -0,0 +1,192 @@
+
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        self._last_same_trap: dict[Tuple[str, int], datetime] = {}
+
+    def handle_trap(self, trap: TrapMessage) -> Optional[bool]:
+        _logger.debug("%s: %s (vars: %s)", trap.agent.device.name, trap.name, ", ".join(trap.variables))
+
+        if "ifIndex" in trap.variables:
+            ifindex = trap.variables.get("ifIndex").value
+        else:
+            _logger.warning("%s: %s trap contained no ifIndex value, ignoring", trap.agent.device.name, trap.name)
+            return False
+        port = trap.agent.device.ports.get(ifindex) if ifindex > 0 else None
+        if not port:
+            _logger.warning(
+                "%s: %s trap referenced unknown port (ix %s), ignoring", trap.agent.device.name, trap.name, ifindex
+            )
+            return False
+
+        # TODO: The trap *might* contain an ifDescr value.  If present, Zino uses that for trap processing.
+        #  Otherwise, it fetches ifDescr from its own state and uses that for trap processing.  Either way,
+        #  there seems to be some redundancy. We should document why, or change the behavior in Zino 2
+
+        # The legacy Zino also looks for `locIfDescr` in the trap at this point, but this is an ancient
+        # Cisco-specific variable we no longer see.  It might have been replaced by `cieIfOperStatusCause`,
+        # but we have no more Cisco devices to test on, so this has been deliberately left out.
+
+        is_up = trap.name == "linkUp"
+        self.handle_link_transition(trap.agent.device, port, is_up)
+
+    def handle_link_transition(self, device: DeviceState, port: Port, is_up: bool, reason: Optional[str] = None):
+        """This method is called when a linkUp or linkDown trap is received, and updates event state accordingly -
+        including the logic for handling flapping states.
+        """
+        if self.is_port_ignored_by_patterns(device, port.ifdescr):
+            return
+
+        # At this point, legacy Zino would ignore so-called "insignificant" interfaces.  As mentioned in the
+        # LinkStateTask docstring, this concept has deliberately not been ported to Zino 2.
+
+        _logger.debug(
+            "%s: intf %s ix %d link%s%s",
+            device.name,
+            port.ifdescr,
+            port.ifindex,
+            "Up" if is_up else "Down",
+            f", {reason}" if reason else "",
+        )
+
+        if self.is_port_ignored_by_policy(device, port, is_up):
+            return
+
+        index = (device.name, port.ifindex)
+        self.update_interface_flapping_score(index)
+
+        new_state = InterfaceState.UP if is_up else InterfaceState.DOWN
+
+        if self.is_interface_flapping(index):
+            # TODO: if event doesn't exist, create it
+            # TODO: Record new number of flaps in event
+            # TODO: When flapcount modulo 100 is zero, log a message with flapping stats
+            pass
```

Comment:
I imagine whatever will be in here should also be separated into its own function